### PR TITLE
feat(`config`): set default evm version to cancun

### LINKED ISSUE #5513: `vm.createSelectFork` should set warm coinbase(EIP-3651)
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (8e365be 2023-07-18T00:28:12.648082999Z)

### What command(s) is the bug in?

_No response_

### Operating System

Linux

### Describe the bug

As [EIP-3651](https://eips.ethereum.org/EIPS/eip-3651), which is applied in Shanghai upgrade, the block.coinbase address should be warm in this block. 
When using the cheatcode createSelectFork, user expect the behaviour to be identical to execute the tx at the top of the specific block. This will have a little effect on gas calculation.

### LINKED ISSUE #7014: bug: `forge config --json` returns evm version that is not compatible with solc version
### Component

Forge, Chisel

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (84d9842 2024-02-02T00:19:34.098872000Z)

### What command(s) is the bug in?

`forge config --json`

### Operating System

macOS (Apple Silicon)

### Describe the bug

When the EVM version isn't set in the foundry.toml but the solc version is, Foundry may incorrectly use `EvmVersion::Paris` without validating that it is available for the given solc verison e.g. solc 0.8.17. I would expect to be able to pass the output of `forge config --json` to solc without running into errors related solc configurations. I'd also expect incompatibilities that are explicitly defined in the foundry.tom to be reported in a diagnostic.

I looked into fixing this but I'm not familiar with the figment provider library and wasn't sure how to add a call to [`normalize_version`](https://github.com/foundry-rs/compilers/blob/9d91a56dadfe0f86d21c875f306e7150098ebac2/src/artifacts/mod.rs#L810) that will produce a correct solc version and evm version combo cleanly (I assume someone with more knowledge of the toolchain and architecture can consolidate the validation and error handling better than myself).

Relevant code for `forge config`
https://github.com/foundry-rs/foundry/blob/7922fd5482f9561699e0fe5a903c90b3fa1fc50d/crates/config/src/lib.rs#L1774
Relevant code for `chisel` which I believe also performs insufficient validation i.e it can produce other misconfigurations for solc version and evm version pairs besides 0.8.17 and Paris.
https://github.com/foundry-rs/foundry/blob/7922fd5482f9561699e0fe5a903c90b3fa1fc50d/crates/chisel/src/session_source.rs#L108-L121

Also, probably [here](https://github.com/foundry-rs/foundry/blob/7922fd5482f9561699e0fe5a903c90b3fa1fc50d/crates/cli/src/utils/cmd.rs#L246) and [here](https://github.com/foundry-rs/foundry/blob/7922fd5482f9561699e0fe5a903c90b3fa1fc50d/crates/config/src/lib.rs#L1024-L1040) could use the same fix.

For more info, see also https://github.com/crytic/slither/issues/2287

### LINKED ISSUE #8614: feat(`config`): upgrade default `evm_version` to `cancun`
### Component

Other (please describe)

### Describe the feature you would like

We should evaluate whether it already makes sense to default to either: `cancun` (13 Mar 2024) or `shanghai` (12 Apr 2023)  from `paris` (15 Sep 2022). ([source](https://docs.blockscout.com/for-developers/information-and-settings/evm-version-information))

The current configuration is:

```toml
evm_version = "paris"
```

The table of https://www.evmdiff.com/features?feature=opcodes shows [`tstore`](https://www.evmdiff.com/features?feature=opcodes#TSTORE) and [`tload`](https://www.evmdiff.com/features?feature=opcodes#TLOAD) are not supported on Avalance and Linea but is available on the other rollups. The current support of the entire `cancun` upgrade on L2s seems largely there, other EVM compatible chains have less support.

For `1.0` I think it also makes sense to establish an expectation that the default we set trails by `x` months.

### Additional context

Considering this is a breaking change and most users have possibly not specified an `evm_version` this change can introduce unintended side effects.

It may make sense to pin to `paris` for `1.0` and apply the breaking changes post-release.

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation
Closes #8614 
Closes #7014  : with updated version we can proper normalize versions post paris
Related tests https://github.com/foundry-rs/foundry/blob/0acbb7249907910aef23f3bff4276481ee62863d/crates/forge/tests/cli/config.rs#L761-L769
Closes #5513 : with updated cancun version, warm coinbase(EIP-3651) is set by default

## Solution

- moved cancun test profile as default. created a paris profile for specific tests
- updated gas and bytecode related test. executed existing gas snapshot related tests with paris to preserve values
- updated forked tests that are not specific to hardforks to use newer blocks/txes
- moved beforeTest with selfdestruct from repros to core and executed with Paris
- changed testEtch to use different address than precompile

### CODE REVIEW

**Comment by zerosnacks** [comment]
Related: https://github.com/foundry-rs/foundry/issues/8614

File: crates/forge/tests/cli/script.rs [hunk 1]
```diff
@@ -242,10 +242,10 @@ Script ran successfully.
 ==========================
 Simulated On-chain Traces:
 
-  [45299] → new GasWaster@[..]
-    └─ ← [Return] 226 bytes of code
+  [44291] → new GasWaster@0x810090f35DFA6B18b5EB59d298e2A2443a2811E2
```
Comment (zerosnacks) [inline crates/forge/tests/cli/script.rs:245]: `[..]` is preferred here I think

File: crates/forge/tests/cli/script.rs [hunk 2]
```diff
@@ -349,10 +349,10 @@ Script ran successfully.
...
+    └─ ← [Return] 221 bytes of code
```
Comment (zerosnacks) [inline crates/forge/tests/cli/script.rs:353]: Same here

File: crates/forge/tests/it/fuzz.rs [hunk 3]
```diff
@@ -203,10 +203,10 @@ contract FuzzerDictTest is Test {
     .unwrap();
 
     // Test that immutable address is used as fuzzed input, causing test to fail.
-    cmd.args(["test", "--fuzz-seed", "100", "--mt", "testImmutableOwner"]).assert_failure();
+    cmd.args(["test", "--fuzz-seed", "119", "--mt", "testImmutableOwner"]).assert_failure();
```
Comment (zerosnacks) [inline crates/forge/tests/it/fuzz.rs:206]: Any reason for this specific change?

File: crates/forge/tests/it/spec.rs [hunk 4]
```diff
@@ -1,14 +1,10 @@
...
 
 #[tokio::test(flavor = "multi_thread")]
 async fn test_shanghai_compat() {
     let filter = Filter::new("", "ShanghaiCompat", ".*spec");
-    TestConfig::with_filter(TEST_DATA_DEFAULT.runner(), filter)
-        .evm_spec(SpecId::SHANGHAI)
-        .run()
-        .await;
+    TestConfig::with_filter(TEST_DATA_DEFAULT.runner(), filter).run().await;
 }
```
Comment (zerosnacks) [inline crates/forge/tests/it/spec.rs:10]: Is this correct? It appears to test specifically against Shanghai compatibility - not sure of the context, perhaps we can generalize the name into `test_compat()`?

File: testdata/default/cheats/Etch.t.sol [hunk 5]
```diff
@@ -8,7 +8,7 @@ contract EtchTest is DSTest {
     Vm constant vm = Vm(HEVM_ADDRESS);
 
     function testEtch() public {
-        address target = address(10);
+        address target = address(11);
```
Comment (zerosnacks) [inline testdata/default/cheats/Etch.t.sol:11]: Is this related to an new precompile in 0x10?

**Review by zerosnacks — COMMENTED** [review]
Supportive of moving to Cancun as default now, some minor notes / questions in regards to changes

File: crates/forge/tests/cli/script.rs (hunk 1 above)
Comment (grandizzy) [reply crates/forge/tests/cli/script.rs:245]: I can revert but this was automatically replaced by running with `SNAPSHOTS=overwrite` and is handy IMO

File: testdata/default/cheats/Etch.t.sol (hunk 5 above)
Comment (grandizzy) [reply testdata/default/cheats/Etch.t.sol:11]: yep, etch cheatcode has `ensure_not_precompile` and cancun added the new one

File: crates/forge/tests/it/fuzz.rs (hunk 3 above)
Comment (grandizzy) [reply crates/forge/tests/it/fuzz.rs:206]: looks like contract bytecode has been changed and it no longer fails in configured runs, changing the seed keeps the runs to catch it but can also bump the runs with prev seed

File: crates/forge/tests/it/spec.rs (hunk 4 above)
Comment (grandizzy) [reply crates/forge/tests/it/spec.rs]: good point, I kept it the way it was, with deafult paris and making sure shanghai compatible, see https://github.com/foundry-rs/foundry/pull/9131/commits/702c270be810314cd81e0a12d19a52e9a934cff0

**Review by zerosnacks — APPROVED** [review]
Lgtm 👍

**Comment by grandizzy** [comment]
I think going forward the test structure and such updates could be simplified a lot with https://github.com/foundry-rs/foundry/issues/8564 that is allowing inline evm config version.

File: crates/forge/tests/cli/script.rs (hunk 1 above)
Comment (grandizzy) [reply crates/forge/tests/cli/script.rs]: :man_facepalming: yes, indeed, better `[..]`, changed in https://github.com/foundry-rs/foundry/pull/9131/commits/1ecd4d03b792ccae4aa771a8e8c063a5d26cf03f

**Comment by zerosnacks** [comment]
pending review cc @yash-atreya / @klkvr

**Review by mattsse — APPROVED** [review]
lgtm
